### DISCUSSION TITLE
How to mount disk image created by multipass to recover data

### DISCUSSION THREAD

**Discussion by mores** [canonical/multipass#3688] [Q&A] [answered]
I have used multipass to stand up ubuntu instance very fast on macos.
But at some time later the image no longer boots.
How can I mount the disk image so I can recover data that I saved in that image that is no longer bootable ?

**Comment by andrei-toterman** [comment, marked as answer]
Hey, @mores! Sorry to hear that you lost access to the files. If you want to investigate the root cause further, I suggest you open an issue in this repository, with as many details as possible that could help us figure out why it happened. Nevertheless, here is how you can recover the files from an instance:

- Make a copy of the image of the VM (repalce `<instance-name>` with the actual name of the instance):
  ```
  sudo cp /var/root/Library/Application\ Support/multipassd/qemu/vault/instances/<instance-name>/*.img ~
  ```
- Now you need access to a Linux machine. This can be just another functional Multipass instance that has enough disk space to be able to contain that `img` file.
- Copy that `img` file to that Linux machine. If you're using another multipass instance, you can use `multipass transfer` to copy it.
- Open a shell on that Linux machine (`multipass shell <instance>` if you're using another instance).
- Now, inside that Linux shell, do the following:
  - Install qemu tools: `sudo apt install qemu-utils`
  - Enable NBD support: `sudo modprobe nbd max_part=16`
  - Connect the image you copied to an nbd device (replace `<image-file>` with the actual file name): 
    `qemu-nbd --connect=/dev/nbd0 <image-file>.img`
  - Mount that NBD device into a new directory: `mkdir instance-files && sudo mount /dev/nbd0p1 instance-files`

Now inside that directory, you should be able to see all the files that were in the instance. I hope this helps. If you encounter any difficulties or something is not clear, let me know!

**Reply by mores** [reply]
Thank you **VERY MUCH** for this info !

In the past I have used VMware and Virtual Box and never any issues with those programs.
multipass did allow me to standup a vm instance on my mac very quickly.
My first instance I tried to increase the size of disk - did not succeed, so I started up a second instance.
The second instance worked for some time, time passed, upgrade happened on my mac, at some point the second instance stopped booting.
I create a third instance, worked for some time, time passed, upgrade happened on my mac, at some point the third instance stopped booting.
I started researching many of the issues/discussions and found out mac networking is very finicky. 
I think someone needs to publish a new tool/shell script that will check for all known issues and help identify why an instance stopped booting.

**Reply by giuliomoro** [reply]
> Mount that NBD device into a new directory: mkdir instance-files && sudo mount /nbd0p1 instance-files

Should that be `sudo mount /dev/ndb0p1` ?

**Reply by andrei-toterman** [reply]
Yes! Thanks for catching that mistake! I've updated my comment.
